src/gradient.toit

```diff
@@ -0,0 +1,303 @@
+// Copyright (C) 2023 Toitware ApS. All rights reserved.
+// Use of this source code is governed by an MIT-style license that can be
+// found in the LICENSE file.
+
+import bitmap show *
+import math
+
+import .common
+import .style
+
+class GradientSpecifier:
+  color/int
+  percent/int
+
+  constructor --.color/int .percent/int:
+
+/**
+Gradients are similar to CSS linear gradients and SVG gradients.
+They are given a list of $GradientSpecifier, each of which has a color and
+  a percentage, indicating where in the gradient the color should appear.
+  The specifiers should be ordered in increasing order of percentage.
+Angles are as in CSS, with 0 degrees being up and 90 degrees being to the right
+  (this is different from text orientations, which go anti-clockwise).
+See https://cssgradient.io/ for a visual explanation and playground for CSS
+  gradients.
+Example:
+```
+  gradient := GradientBackground --angle=45
+      --specifiers=[
+          GradientSpecifier --color=0xff0000 10,    // Red from 0-10%, red-to-green from 10-50%.
+          GradientSpecifier --color=0x00ff00 50,    // Green-to-blue from 50-90%.
+          GradientSpecifier --color=0x0000ff 90,    // Blue from 90-100%.
+      ]
+```
+*/
+class GradientBackground implements Background:
+  angle/int
+  specifiers/List
+  hash_/int? := null
+  rendering_/GradientRendering_? := null
+
+  constructor --angle/int --.specifiers/List:
+    this.angle = normalize_angle_ angle
+    validate_specifiers_ specifiers
+
+  operator == other -> bool:
+    if other is not GradientBackground: return false
+    if other.angle != angle: return false
+    if other.specifiers.size != specifiers.size: return false
+    for i := 0; i < specifiers.size; i++:
+      if other.specifiers[i].color != specifiers[i].color: return false
+      if other.specifiers[i].percent != specifiers[i].percent: return false
+    return true
+
+  hash_code -> int:
+    if not hash_:
+      hash := 0
+      specifiers.do: | it |
+        hash = (hash * 31) & 0xfff_ffff
+        hash += it.color + 47 * it.percent
+        hash_ = hash
+    return hash_
+
+  draw canvas/Canvas x/int y/int w/int h/int --autocropped/bool -> none:
+    if not rendering_: rendering_ = GradientRendering_.get w h this
+    rendering_.draw canvas x y --autocropped=autocropped
+
+  static normalize_angle_ angle/int -> int:
+    if 0 <= angle < 360:
+      return angle
+    else if angle < 0:
+      return angle % 360 + 360
+    else:
+      return angle % 360
```

Comment:
Does pretty much the same, but doesn't rely on any modulo implementation (and should thus be easier to read).
I'm fine eitherway.
```suggestion
    angle %= 360
    if angle < 0:
      return angle + 360
    return angle
```

Reply:
Done.